powerauth-restful-security-spring/src/main/java/io/getlime/security/powerauth/rest/api/spring/service/v3/UserInfoService.java

```diff
@@ -0,0 +1,109 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.getlime.security.powerauth.rest.api.spring.service.v3;
+
+import com.wultra.security.powerauth.client.PowerAuthClient;
+import com.wultra.security.powerauth.client.model.error.PowerAuthClientException;
+import com.wultra.security.powerauth.client.v3.ActivationStatus;
+import com.wultra.security.powerauth.client.v3.GetActivationStatusResponse;
+import io.getlime.security.powerauth.rest.api.model.entity.UserInfoStage;
+import io.getlime.security.powerauth.rest.api.spring.exception.PowerAuthUserInfoException;
+import io.getlime.security.powerauth.rest.api.spring.provider.UserInfoProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.time.Instant;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Service for obtaining user info as claims map.
+ *
+ * @author Petr Dvorak, petr@wultra.com
+ */
+@Service
+public class UserInfoService {
+
+    private static final Logger logger = LoggerFactory.getLogger(UserInfoService.class);
+
+    private UserInfoProvider userInfoProvider;
+    private final PowerAuthClient powerAuthClient;
+
+    @Autowired
+    public UserInfoService(PowerAuthClient powerAuthClient) {
+        this.powerAuthClient = powerAuthClient;
+    }
+
+    /**
+     * Setter with optional user info provider bean.
+     * @param userInfoProvider User info provider.
+     */
+    @Autowired(required = false)
+    public void setActivationProvider(UserInfoProvider userInfoProvider) {
+        this.userInfoProvider = userInfoProvider;
+    }
+
+    /**
+     * Fetch user info as a map of claims.
+     *
+     * @param activationId Activation ID.
+     * @return Map of claims.
+     * @throws PowerAuthUserInfoException In case there is an error while fetching claims.
+     */
+    public Map<String, Object> fetchUserClaimsByActivationId(String activationId) throws PowerAuthUserInfoException {
+        try {
+            // Fetch activation details
+            final GetActivationStatusResponse activationStatusResponse = powerAuthClient.getActivationStatus(activationId);
+            final String userId = activationStatusResponse.getUserId();
+            final String applicationId = activationStatusResponse.getApplicationId();
+            final ActivationStatus activationStatus = activationStatusResponse.getActivationStatus();
+
+            if (ActivationStatus.ACTIVE != activationStatus) { // only allow active state for now
+                throw new PowerAuthUserInfoException("Invalid activation status: " + activationStatus + ", for activation: " + activationId);
+            }
+
+            final Map<String, Object> map = new LinkedHashMap<>();
+            map.put("sub", userId);
```

Comment:
`sub` may be overridden by `userInfoProvider`, is it desirable?

Reply:
I was thinking, as this is our user-info we should return our userId. So from my POV this is fine.

Reply:
Most important here is that we should NOT provide this internal user identifier by default to the mobile client. If so, then this has to be configurable (by the provider for example) and by default turned off.

Reply:
I spent some time thinking about this while coding - the reasoning is a bit less "virtual and abstract" when actually writing the code. What should this new standard PA RESTful API endpoint return by default?

**Current Status: Minimal claims.** The implementation returns `sub`, `jti` and `iat`, which is the minimal info about the user. The implementor can override the values of these claims, which is a desired behavior documented in JavaDoc. The idea is to allow anonymization of user ID if needed.

**Option 1: Empty object.** It is possible but sure, not very useful. The endpoint would return no claim the client app can rely on.

**Option 2: Anonymous claims.** We can return only `jti` (JWT ID) and `iat` (Issued At timestamp) claim, but what is this good for? The values do not make much sense.

**Option 3: Minimal claims, revised.** Instead of putting the user ID in the `sub` claim, we can add the activation ID as default, which is slightly less problematic and less useless than Option 2... but still pretty useless.

**Option 4: The endpoint itself is conditional.** This does not make much sense either because a standard client in the SDK would need to know if the endpoint is published or not for it to be able to call it.

What is the issue in obtaining the user ID value in the `sub` claim anyway? I think the big issue would be storing the value locally on the device but getting it in the memory inside a personalized device (according to existing activation) seems less problematic...

Comments? Discussion? Ideas? @hvge @romanstrobl @banterCZ 

_Note: We prioritize ease of deployment._

Reply:
I would make the behaviour configurable (i.e. return safe value by default but make it easy to return userId mapping), behaviour configured by a property. This way we can have ease of deployment (using configuration property - this can be automated if needed) and safety by default.

Reply:
🤔 I can add a provider method `shouldAddMinimalStandardClaims(...)` that would return false by default... but what do we put in the vanilla Enrollment Server? Isn't it a customization we will always override? ...

Reply:
Btw this would be simple to deploy on server, but useless for the client. The mobile SDK will have a method `gimmeClaims` that would return an empty map by default. This does not sound very helpful.

Reply:
I think this is a topic for a bigger discussion about the Enrollment Server architecture for client customizations. 

How do we want to keep overriding Enrollment Server functionality in the future? I think the forks are honestly a horrible way (too much work to sync all the downstreams, too little code changes, too many repos, artifact name issues, access rights issues, ...). We can probably come up with something better. 

I would prefer a world, in which:
- vanilla enrollment server is highly configurable - several typical options are implemented for standard activation customization, for user info customization, configuration of activation code handler and for other stuff we usually override using code in a downstream with sensible defaults
- additional code changes can be done externally (e.g. using a library which is added to the enrollment server on build and can contain a much bigger override of enrollment server functionality via specific providers and can even add new endpoints) - we can probably figure out the client code separation, it's just a matter of building to make sure the right code is compiled and included
 - there are many ways how to achieve this - compile time, build time, jar loaded from a database, dynamic language compilation e.g. Groovy, OSGi, ... IMO any of these options is better than a fork

Once we decide about the two topics above, I think the decisions about default behaviour will be slightly different, because the default will become both super easily configurable and super easily overridable.

Reply:
These are good topics to think about, but I think unrelated to the problem and enrollment server. The question is much simpler: What should the standard `/pa/user/info` endpoint return to provide a value to the client?

Reply:
I think the most useful is the `userId`, for sure. However, should this be the default? Hmm.. now we are getting into dangerous waters. What is the process to disable sending the `userId`? Create a fork which disables it? 

I like the option of making the endpoint conditional and return `userId` by default, because this requires some effort and won't be done by mistake. However in this case we should return to the mobile application information about availability of user info endpoint using configuration endpoint.

We should probably discuss it. 

Anyway I think the bigger picture I outlined is also important, because it influences e.g. whether we can easily create templates with default configuration, so we could have an ultra safe setting template and most usable setting template depending on the client profile (e.g. bank vs. fintech). Right now we are choosing the best setting for everyone, which is why it is difficult.

Reply:
To my original question
> `sub` may be overridden by `userInfoProvider`

I see the answer in javadoc

> If these claims are set, they will override values which were automatically inferred from the authentication object provided by PowerAuth stack.

To the Roman's question

> What is the process to disable sending of the userId

I guess that the answer is to implement `UserInfoProvider` providing whatever `sub` to override it. Am I correct?

Downstreams are still painful, but with separated logic, it is a little bit easier to avoid merge conflicts.